tools/shared.py

```diff
@@ -2730,16 +2730,21 @@ def clang_preprocess(filename):
   return run_process([CLANG_CC, '-DFETCH_DEBUG=1', '-E', '-P', '-C', '-x', 'c', filename], check=True, stdout=subprocess.PIPE).stdout
 
 def read_and_preprocess(filename):
-  f = open(filename, 'r').read()
-  pos = 0
-  include_pattern = re.compile('^#include\s*["<](.*)[">]\s?$', re.MULTILINE)
-  while(1):
-    m = include_pattern.search(f, pos)
-    if not m:
-      return f
-    included_file = open(os.path.join(os.path.dirname(filename), m.groups(0)[0]), 'r').read()
+  temp_dir = configuration.get_temp_files().get_dir()
+
+  # Create a settings file with the current settings to pass to the JS preprocessor
+  settings_str = "var " + ";\nvar ".join(Settings.serialize()[1::2])
+  settings_file = os.path.join(temp_dir, 'settings.js')
+  open(settings_file, 'w').write(settings_str)
 
-    f = f[:m.start(0)] + included_file + f[m.end(0):]
+  # Run the JS preprocessor
+  (path, file) = os.path.split(filename)
+  args=[settings_file, file]
+  stdout = os.path.join(temp_dir, 'stdout')
+  run_js(path_from_root('tools/preprocessor.js'), NODE_JS, args, True, stdout=open(stdout, 'w'), cwd=path)
```

Comment:
do you need a temp file for stdout? i think you can use `PIPE` and just get the output from `run_js`?

Reply:
(and in that case, i don't think `temp_dir` is needed, can just get a temp file for the settings file directly)